projects/kit/directives/highlight/highlight.directive.ts

```diff
@@ -12,90 +39,162 @@ import {Observable} from 'rxjs';
-                continue;
+            if (!this.tuiHighlightMultiOccurrences) {
+                return;
             }
+        }
+    }
 
-            const range = this.doc.createRange();
+    private createHighlight(
+        hostRect: DOMRect,
+        {left, top, width, height}: DOMRect,
+    ): Observable<never> {
+        return new Observable<never>(() => {
+            const ref = this.vcr.createComponent(this.cf);
+            const {instance} = ref;
 
-            range.setStart(this.treeWalker.currentNode, index);
-            range.setEnd(this.treeWalker.currentNode, index + this.tuiHighlight.length);
+            instance.left = left - hostRect.left;
+            instance.top = top - hostRect.top;
+            instance.width = width;
+            instance.height = height;
 
-            const hostRect = this.el.nativeElement.getBoundingClientRect();
-            const {left, top, width, height} = range.getBoundingClientRect();
-            const {style} = this.highlight;
+            this.el.nativeElement.appendChild(ref.location.nativeElement);
 
-            style.background = this.tuiHighlightColor;
-            style.left = tuiPx(left - hostRect.left);
-            style.top = tuiPx(top - hostRect.top);
-            style.width = tuiPx(width);
-            style.height = tuiPx(height);
-            style.display = 'block';
+            ref.changeDetectorRef.detectChanges();
 
-            return;
-        } while (this.treeWalker.nextNode());
+            return () => ref.destroy();
+        });
     }
 
-    private indexOf(source: string | null): number {
-        return !source || !this.tuiHighlight
-            ? -1
-            : source.toLowerCase().indexOf(this.tuiHighlight.toLowerCase());
+    private *getOccurrences(source: string | null): Generator<TuiHighlightOccurrence> {
+        if (!source || !this.tuiHighlight) {
+            return;
+        }
+
+        for (const item of tuiToArray(this.tuiHighlight)) {
+            if (tuiIsString(item)) {
+                const itemValue = this.tuiHighlightCaseSensitive
+                    ? item
+                    : item.toLowerCase();
+                const sourceValue = this.tuiHighlightCaseSensitive
+                    ? source
+                    : source.toLowerCase();
+
+                for (
+                    let index = sourceValue.indexOf(itemValue);
+                    index >= 0;
+                    index = sourceValue.indexOf(itemValue, index + 1)
+                ) {
+                    yield {
+                        index,
+                        length: itemValue.length,
+                    };
+                }
+            } else {
+                for (const match of source.matchAll(item)) {
+                    if (tuiIsNumber(match.index) && match.length) {
+                        yield {
+                            index: match.index,
+                            length: match[0].length,
+                        };
```

Comment:
What are the benefits of using generators?

Reply:
1. We can control how many occurrences we want to calculate. In `match` getter, for example, we need only one occurrence for say "yes". In second, if we pass tuiHighlightMultiOccurrences: false, then also we need to calculate only one occurrence.
2. In generators we can split iteration into smaller parts of logic for cleaner code, without iterate same array several times. We can execute all calculations on the fly.